### PR TITLE
Connect bindings to locals and calls to functions

### PR DESCRIPTION
**Pull request by briansteffens**

### CODE REVIEW

File: ast.d [hunk 1]
```diff
@@ -148,6 +150,30 @@ OperatorType parseOperatorType(string input)
     }
 }
 
+enum OperatorClass
+{
+    Math,
+    Relational,
+    Logical
+}
```
Comment (eatonphil) [inline ast.d:163]: What are these for?

Comment (briansteffens) [reply ast.d:163]: For areas like this https://github.com/briansteffens/bshift/pull/6/files#diff-0a5963e354b6ffca029f3be2efa243a8R192

Comment (eatonphil) [reply ast.d:163]: Why do the right / left types only matter for the Math operators? Also, is string concatenation a "Math" operator?

Comment (eatonphil) [reply ast.d:163]: Or is the plan just to support C-like levels of operators?

Comment (briansteffens) [reply ast.d:163]: ```==```, ```!=```, ```&&```, ```||```, etc all return bools no matter what their inputs are right? Haven't thought about string concatenation yet.

Comment (briansteffens) [reply ast.d:163]: D uses ```~``` for string/array concatenation, which is pretty nice for being more clear about what operation is happening.

Comment (eatonphil) [reply ast.d:163]: Is there a need to distinguish between Relational and Logical operators?

Comment (briansteffens) [reply ast.d:163]: Could probably make the generate functions for both types render the same assembly but relational operators can get away with no jumps, while logical need jumps to allow stopping evaluation early when the answer is known from the left side only.

**Comment by eatonphil** [comment]
> Yeah it was rendering [rbp--1] when it got here lol

Ah, lol

**Comment by briansteffens** [comment]
Function name mangling is gone as of last commit. Also there's a new preParse phase that just looks at imports, function definitions, and externs so that the real parse can hook up calls to their signatures and set node types appropriately.

File: test/auto_from_array.asmtest [hunk 2]
```diff
@@ -0,0 +1,21 @@
+u64 main()
+{
+    u64[2] ar;
+    ar[0] = 3;
+    ar[1] = 7;
+
+    auto ar2 = ar;
```
Comment (eatonphil) [inline test/auto_from_array.asmtest]: That's pretty cool! Or maybe it's simpler than it seems.

File: test/auto_from_array_index.asmtest [hunk 3]
```diff
@@ -0,0 +1,20 @@
+u64 main()
+{
+    u64[2] ar;
+    ar[0] = 3;
+    ar[1] = 7;
+
+    auto x = ar[0];
```
Comment (eatonphil) [inline test/auto_from_array_index.asmtest]: This is also pretty cool!

**Pull request merged**